### PR TITLE
[DUBBO-4010] TypeDefinitonBuilder support google protobuf entity build #

### PR DESCRIPTION
**Pull request by vio-lin**
## What is the purpose of the change
TypeDefinitonBuilder support google protobuf service metadata generating

## Brief changelog
1. change TypeDefinitionBuilder  use SPI to load custom TypeBuilder
2. add GooglePbTypeBuilder in dubbo-serialization-protobuf-json to build Protobuf entity class

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/incubator-dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-metadata-report/dubbo-metadata-definition/src/main/java/org/apache/dubbo/metadata/definition/builder/MapTypeBuilder.java [hunk 1]
```diff
@@ -58,11 +59,7 @@ public TypeDefinition build(Type type, Class<?> clazz, Map<Class<?>, TypeDefinit
                 TypeDefinitionBuilder.build(actualType, rawType, typeCache);
             } else if (actualType instanceof Class<?>) {
                 Class<?> actualClass = (Class<?>) actualType;
-                if (actualClass.isArray() || actualClass.isEnum()) {
```
Comment (beiwei30) [inline dubbo-metadata-report/dubbo-metadata-definition/src/main/java/org/apache/dubbo/metadata/definition/builder/MapTypeBuilder.java]: any reason to remove this?

File: dubbo-metadata-report/dubbo-metadata-definition/src/main/java/org/apache/dubbo/metadata/definition/builder/CollectionTypeBuilder.java [hunk 2]
```diff
@@ -60,11 +61,7 @@ public TypeDefinition build(Type type, Class<?> clazz, Map<Class<?>, TypeDefinit
             TypeDefinitionBuilder.build(actualType, rawType, typeCache);
         } else if (actualType instanceof Class<?>) {
             Class<?> actualClass = (Class<?>) actualType;
-            if (actualClass.isArray() || actualClass.isEnum()) {
```
Comment (beiwei30) [inline dubbo-metadata-report/dubbo-metadata-definition/src/main/java/org/apache/dubbo/metadata/definition/builder/CollectionTypeBuilder.java]: any reason to remove this?

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+import java.lang.reflect.Method;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ProtobufTypeBuilder implements TypeBuilder {
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: pls. change to 'private'

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+        }
+    }
+
+    @Override
+    public boolean accept(Type type, Class<?> clazz) {
+        if (clazz == null) {
+            return false;
+        }
+
+        if (GeneratedMessageV3.class.isAssignableFrom(clazz)) {
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: pls. simplify it to 'return GeneratedMessageV3.class.isAssignableFrom(clazz);'

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+            return true;
+        }
+
+        return false;
+    }
+
+    @Override
+    public TypeDefinition build(Type type, Class<?> clazz, Map<Class<?>, TypeDefinition> typeCache) {
+        TypeDefinition typeDefinition = new TypeDefinition(clazz.getName());
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: shouldn't it be 'TypeDefinition typeDefinition = null'?

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+        return (GeneratedMessageV3.Builder) method.invoke(null, null);
+    }
+
+    private TypeDefinition buildProtobufTypeDefinition(Class<?> clazz, GeneratedMessageV3.Builder builder, Map<Class<?>, TypeDefinition> typeCache) {
+        TypeDefinition typeDefinition = new TypeDefinition(clazz.getName());
+        if (builder == null) {
+            return typeDefinition;
+        }
+
+        Map<String, TypeDefinition> properties = new HashMap();
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: should be 'new HashMap<>()'.

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+        Method[] methods = builder.getClass().getDeclaredMethods();
+        for (Method method : methods) {
+            String methodName = method.getName();
+
+            if (isSimplePropertySettingMethod(method)) {
+                // property of custom type or primitive type
+                properties.put(generateSimpleFiledName(methodName), TypeDefinitionBuilder.build(method.getGenericParameterTypes()[0], method.getParameterTypes()[0], typeCache));
+                continue;
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: continue is unnecessary

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+            } else if (isMapPropertySettingMethod(method)) {
+                // property of map
+                Type type = method.getGenericParameterTypes()[0];
+                String fieldName = generateMapFieldName(methodName);
+                validateMapType(fieldName, type.toString());
+                properties.put(fieldName, TypeDefinitionBuilder.build(type, method.getParameterTypes()[0], typeCache));
+                continue;
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: continue is unnecessary

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+                TypeDefinition td;
+                if (ProtocolStringList.class.isAssignableFrom(method.getReturnType())) {
+                    // property defined as "repeated string" transform to ProtocolStringList,
+                    // should be build as List<String>.
+                    td = TypeDefinitionBuilder.build(STRING_LIST_TYPE, List.class, typeCache);
+                } else {
+                    td = TypeDefinitionBuilder.build(type, method.getReturnType(), typeCache);
+                }
+                properties.put(fieldName, td);
+                continue;
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: continue is unnecessary

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class ProtobufTypeBuilder implements TypeBuilder {
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+    private static final Pattern MAP_PATTERN = Pattern.compile("^java[.]util[.]Map<(\\w+[[.$]\\w*]+), (\\w+[[.$]\\w*]+)>$");
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: is this pattern too strict and unnecessary?  is this better? "^java[.]util[.]Map<(\\S+), (\\S+)>$")"

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+    private static final Pattern LIST_PATTERN = Pattern.compile("^java[.]util[.]List<(\\w+([.$]\\w*)+)>$");
```
Comment (beiwei30) [inline dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: ditto

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 3 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 4 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 6 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 7 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 8 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 9 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 10 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: it looks more clean. done

File: dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java (hunk 11 above)
Comment (vio-lin) [reply dubbo-serialization/dubbo-serialization-protobuf-json/src/main/java/org/apache/dubbo/common/serialize/protobuf/support/ProtobufTypeBuilder.java]: ^java\.util\.List<(\S+)>$
like map Pattern

**Pull request merged**